### PR TITLE
feat: Add InlinePcesWriter

### LINKED ISSUE #15605: Create new inline PCES Writer component
Create a new unwired component that writes events to the PCES according to the proposal. This component should be well tested with unit tests but not wired. As much code should be reused from the existing PCES component as possible. There should not be a flush method in this component (existing component has `submitFlushRequest()` which should not exist in this new component. The return type of `writeEvent()` should return an event.

### PR DESCRIPTION
**Pull request by timo0**
Fix #15605

### CODE REVIEW

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/preconsensus/CommonPcesWriter.java [hunk 1]
```diff
@@ -0,0 +1,414 @@
...
+        }
+    }
+
+    /**
+     * Let the event writer know the current non-ancient event boundary. Ancient events will be ignored if added to the
+     * event writer.
+     *
+     * @param nonAncientBoundary describes the boundary between ancient and non-ancient events
+     */
+    public void updateNonAncientEventBoundary(@NonNull EventWindow nonAncientBoundary) {
```
Comment (lpetrovic05) [inline platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/preconsensus/CommonPcesWriter.java:201]: final

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/preconsensus/CommonPcesWriter.java [hunk 2]
```diff
@@ -0,0 +1,414 @@
...
+    public long getLastFlushedEvent() {
+        return lastFlushedEvent;
+    }
+
+    /**
+     * Get the average span utilization.
+     *
+     * @return the average span utilization
+     */
+    public boolean isBootstrapMode() {
```
Comment (lpetrovic05) [inline platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/preconsensus/CommonPcesWriter.java:380]: method seems to be unused

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/preconsensus/CommonPcesWriter.java [hunk 3]
```diff
@@ -0,0 +1,414 @@
...
+        return bootstrapMode;
+    }
+
+    /**
+     * Set the last written event.
+     *
+     * @param lastWrittenEvent the last written event
+     */
+    public void setLastWrittenEvent(long lastWrittenEvent) {
```
Comment (lpetrovic05) [inline platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/preconsensus/CommonPcesWriter.java:389]: final

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/preconsensus/CommonPcesWriter.java [hunk 4]
```diff
@@ -0,0 +1,414 @@
...
+        this.lastWrittenEvent = lastWrittenEvent;
+    }
+
+    /**
+     * Set the last flushed event.
+     *
+     * @param lastFlushedEvent the last flushed event
+     */
+    public void setLastFlushedEvent(long lastFlushedEvent) {
```
Comment (lpetrovic05) [inline platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/preconsensus/CommonPcesWriter.java:398]: final

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/preconsensus/DefaultInlinePcesWriter.java [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+            return event;
+        }
+
+        // don't do anything with ancient events
+        if (event.getAncientIndicator(commonPcesWriter.getFileType()) < commonPcesWriter.getNonAncientBoundary()) {
+            return null;
+        }
+
+        try {
+            final boolean fileClosed = commonPcesWriter.prepareOutputStream(event);
```
Comment (lpetrovic05) [inline platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/preconsensus/DefaultInlinePcesWriter.java:75]: fileClosed is never checked

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/preconsensus/CommonPcesWriter.java [hunk 6]
```diff
@@ -0,0 +1,414 @@
...
+        }
+
+        // if the block above closed the file, then we need to create a new one
+        if (currentMutableFile == null) {
+            final long upperBound = nonAncientBoundary
+                    + computeNewFileSpan(nonAncientBoundary, eventToWrite.getAncientIndicator(fileType));
+
+            currentMutableFile = fileManager
+                    .getNextFileDescriptor(nonAncientBoundary, upperBound)
+                    .getMutableFile();
```
Comment (lpetrovic05) [inline platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/preconsensus/CommonPcesWriter.java:294]: The main feature of the inline PCES writer should be that it ensures that events are persisted before returning. The way this is currently implemented is by calling `getMutableFile(false, true)`

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/preconsensus/InlinePcesWriter.java [hunk 7]
```diff
@@ -43,7 +43,7 @@ public interface InlinePcesWriter {
      * @return the event written
      */
     @InputWireLabel("events to write")
-    @NonNull
+    @Nullable
```
Comment (lpetrovic05) [inline platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/preconsensus/InlinePcesWriter.java:46]: Why is it nullable? It should not drop any events

**Comment by lpetrovic05** [comment]
I think there should be at least some tests for the new class

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/preconsensus/CommonPcesWriter.java [hunk 8]
```diff
@@ -0,0 +1,414 @@
...
+        }
+        streamingNewEvents = true;
+    }
+
+    /**
+     * Inform the preconsensus event writer that a discontinuity has occurred in the preconsensus event stream.
+     *
+     * @param newOriginRound the round of the state that the new stream will be starting from
+     */
+    public void registerDiscontinuity(@NonNull Long newOriginRound) {
```
[the remark this answers is not in the document]
Comment (timo0) [reply platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/preconsensus/CommonPcesWriter.java]: done, thanks

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/preconsensus/CommonPcesWriter.java (hunk 1 above)
Comment (timo0) [reply platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/preconsensus/CommonPcesWriter.java]: done, thanks

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/preconsensus/CommonPcesWriter.java (hunk 2 above)
Comment (timo0) [reply platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/preconsensus/CommonPcesWriter.java]: removed, thanks

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/preconsensus/CommonPcesWriter.java (hunk 3 above)
Comment (timo0) [reply platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/preconsensus/CommonPcesWriter.java]: done, thanks

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/preconsensus/CommonPcesWriter.java (hunk 4 above)
Comment (timo0) [reply platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/preconsensus/CommonPcesWriter.java]: done, thanks

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/preconsensus/DefaultInlinePcesWriter.java (hunk 5 above)
Comment (timo0) [reply platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/preconsensus/DefaultInlinePcesWriter.java]: variable removed

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/preconsensus/CommonPcesWriter.java (hunk 6 above)
Comment (timo0) [reply platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/preconsensus/CommonPcesWriter.java]: added to the `CommonPcesWriter`

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/preconsensus/InlinePcesWriter.java (hunk 7 above)
Comment (timo0) [reply platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/preconsensus/InlinePcesWriter.java]: reverted

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/preconsensus/DefaultInlinePcesWriter.java [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+    @Override
+    public PlatformEvent writeEvent(@NonNull PlatformEvent event) {
+        if (event.getStreamSequenceNumber() == PlatformEvent.NO_STREAM_SEQUENCE_NUMBER) {
+            throw new IllegalStateException("Event must have a valid stream sequence number");
+        }
+
+        // if we aren't streaming new events yet, assume that the given event is already durable
+        if (!commonPcesWriter.isStreamingNewEvents()) {
+            commonPcesWriter.setLastWrittenEvent(event.getStreamSequenceNumber());
+            commonPcesWriter.setLastFlushedEvent(event.getStreamSequenceNumber());
```
Comment (lpetrovic05) [inline platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/preconsensus/DefaultInlinePcesWriter.java:64]: again, sequence numbers

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/preconsensus/DefaultInlinePcesWriter.java [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+            return event;
+        }
+
+        // don't do anything with ancient events
+        if (event.getAncientIndicator(commonPcesWriter.getFileType()) < commonPcesWriter.getNonAncientBoundary()) {
+            throw new IllegalStateException("Ancient events should not be written to the PCES");
+        }
```
Comment (lpetrovic05) [inline platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/preconsensus/DefaultInlinePcesWriter.java:71]: we shouldn't throw when getting an ancient event, we should just pass it along

**Comment by lpetrovic05** [comment]
I think there isn't any test for the inline PCES writer

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/preconsensus/DefaultInlinePcesWriter.java [hunk 11]
```diff
@@ -54,29 +54,21 @@ public void beginStreamingNewEvents() {
...
+            // don't do anything with ancient events
+            return event;
         }
 
         try {
             commonPcesWriter.prepareOutputStream(event);
             commonPcesWriter.getCurrentMutableFile().writeEvent(event);
-            commonPcesWriter.setLastWrittenEvent(event.getStreamSequenceNumber());
 
             commonPcesWriter.getCurrentMutableFile().flush();
```
Comment (lpetrovic05) [inline platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/preconsensus/DefaultInlinePcesWriter.java:71]: inline writer flushes on every event, so no need for explicit flush

Comment (timo0) [reply platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/preconsensus/DefaultInlinePcesWriter.java]: done

File: platform-sdk/swirlds-platform-core/src/test/java/com/swirlds/platform/event/preconsensus/DefaultInlinePcesWriterTest.java [hunk 12]
```diff
@@ -0,0 +1,137 @@
...
+        final TransactionGenerator transactionGenerator = buildTransactionGenerator();
+
+        return new StandardGraphGenerator(
+                platformContext,
+                random.nextLong(),
+                new StandardEventSource().setTransactionGenerator(transactionGenerator),
+                new StandardEventSource().setTransactionGenerator(transactionGenerator),
+                new StandardEventSource().setTransactionGenerator(transactionGenerator),
+                new StandardEventSource().setTransactionGenerator(transactionGenerator));
+    }
```
Comment (lpetrovic05) [inline platform-sdk/swirlds-platform-core/src/test/java/com/swirlds/platform/event/preconsensus/DefaultInlinePcesWriterTest.java]: We should be able to replace all of this with the `TestingEventBuilder`

**Review by lpetrovic05 — APPROVED** [review]

File: platform-sdk/swirlds-platform-core/src/test/java/com/swirlds/platform/event/preconsensus/DefaultInlinePcesWriterTest.java (hunk 12 above)
Comment (timo0) [reply platform-sdk/swirlds-platform-core/src/test/java/com/swirlds/platform/event/preconsensus/DefaultInlinePcesWriterTest.java]: No there is some issues with BIRTH_ROUND handling with the builder

**Pull request merged**